beacon_chain/validation/src/block_validation.rs

```diff
@@ -0,0 +1,370 @@
+         */
+        let pow_chain_ref = b.pow_chain_ref();
+        if !self.pow_store.block_hash_exists(b.pow_chain_ref())? {
+            return Err(SszBlockValidationError::UnknownPoWChainRef);
+        }
+
+        /*
+         * Store a slice of the serialized attestations from the block SSZ.
+         */
+        let attestations_ssz = &b.attestations();
+
+        /*
+         * Get a slice of the first serialized attestation (the 0'th) and decode it into
+         * a full AttestationRecord object.
+         *
+         * The first attestation must be validated separately as it must contain a signature of the
+         * proposer of the previous block (this is checked later in this function).
+         */
+        let (first_attestation_ssz, next_index) = split_one_attestation(
+            &attestations_ssz,
+            0)?;
+        let (first_attestation, _) = AttestationRecord::ssz_decode(
+            &first_attestation_ssz, 0)?;
+
+        /*
+         * The first attestation may not have oblique hashes.
+         *
+         * The presence of oblique hashes in the first attestation would indicate that the proposer
+         * of the previous block is attesting to some other block than the one they produced.
+         */
+        if first_attestation.oblique_parent_hashes.len() > 0 {
+            return Err(SszBlockValidationError::ProposerAttestationHasObliqueHashes);
+        }
+
+        /*
+         * Generate the context in which attestations will be validated.
+         */
+        let attestation_validation_context = Arc::new(AttestationValidationContext {
+            block_slot,
+            cycle_length: self.cycle_length,
+            last_justified_slot: self.last_justified_slot,
+            parent_hashes: self.parent_hashes.clone(),
+            block_store: self.block_store.clone(),
+            validator_store: self.validator_store.clone(),
+            attester_map: self.attester_map.clone(),
+        });
+
+        /*
+         * Validate this first attestation.
+         */
+        let attestation_voters = attestation_validation_context
+            .validate_attestation(&first_attestation)?;
+
+        /*
+         * Read the parent hash from the block we are validating then attempt to load
+         * the parent block ssz from the database. If that parent doesn't exist in
+         * the database, reject the block.
+         *
+         * If the parent does exist in the database, read the slot of that parent. Then,
+         * determine the proposer of that slot (the parent slot) by looking it up
+         * in the proposer map.
+         *
+         * If that proposer (the proposer of the parent block) was not present in the first (0'th)
+         * attestation of this block, reject the block.
+         */
+        let parent_hash = b.parent_hash();
+        match self.block_store.get_serialized_block(&parent_hash)? {
+            None => return Err(SszBlockValidationError::UnknownParentHash),
```

Comment:
You're right. It would be better to:

1. Load the parent block and learn its slot.
2. Process the attestation.
3. Check that the parent block proposer signed the attestation record.

Attestation validation involves DB reads and will take longer than checking the parent block so there's no good reason to do it first.

I didn't realize I was unnecessarily combining steps 1 & 3.

Good catch, thanks.